File: backend/pipeline/stage09_consistency.py
from typing import List, Set, Dict
from backend.models.schemas import Project

def run_stage09_consistency(project: Project) -> Project:
    sot = project.source_of_truth
    if not sot:
        raise ValueError("Source of Truth missing before Stage 09")

    valid_char_ids: Set[str] = {c.id for c in sot.characters}
    valid_char_names: Dict[str, str] = {c.id: c.name for c in sot.characters}
    valid_loc_ids: Set[str] = {l.id for l in sot.locations}

    total_issues_count = 0

    for scene in project.scenes:
        issues: List[str] = []

        # 1. Verify characters present exist in Source of Truth
        for char_id in scene.characters_present:
            if char_id not in valid_char_ids:
                issues.append(f"Unknown character ID '{char_id}' present in scene.")

        # 2. Verify location ID exists in Source of Truth
        if scene.location_id not in valid_loc_ids:
            issues.append(f"Unknown location ID '{scene.location_id}' assigned to scene.")

        # 3. Verify character names are referenced in image prompt if characters are present
        image_prompt_lower = (scene.image_prompt or "").lower()
        for char_id in scene.characters_present:
            char_name = valid_char_names.get(char_id)
            if char_name:
                first_name = char_name.split()[0].lower()
                if first_name not in image_prompt_lower and char_name.lower() not in image_prompt_lower:
                    issues.append(f"Notice: Character name '{char_name}' not explicitly mentioned in visual prompt text.")

        scene.consistency_issues = issues
        total_issues_count += len(issues)

    project.stage_outputs["stage_09"] = {
        "total_scenes_checked": len(project.scenes),
        "total_issues_flagged": total_issues_count,
        "status": "CLEAN_AND_CONSISTENT" if total_issues_count == 0 else "CONSISTENCY_NOTICES_RECORDED",
    }

    return project

File: server/pipeline/stage09Consistency.ts
import { Project } from '../types.ts';

export async function runStage09Consistency(project: Project): Promise<Project> {
  const sot = project.source_of_truth;
  if (!sot) throw new Error('Source of Truth missing before Stage 09');

  const validCharIds = new Set(sot.characters.map((c) => c.id));
  const validCharNames = new Map(sot.characters.map((c) => [c.id, c.name]));
  const validLocIds = new Set(sot.locations.map((l) => l.id));

  let totalIssuesCount = 0;

  for (const scene of project.scenes) {
    const issues: string[] = [];

    // 1. Verify characters present exist in Source of Truth
    for (const charId of scene.characters_present) {
      if (!validCharIds.has(charId)) {
        issues.push(`Unknown character ID '${charId}' present in scene.`);
      }
    }

    // 2. Verify location ID exists in Source of Truth
    if (!validLocIds.has(scene.location_id)) {
      issues.push(`Unknown location ID '${scene.location_id}' assigned to scene.`);
    }

    // 3. Verify character names are referenced in image prompt if characters are present
    const imagePromptLower = (scene.image_prompt || '').toLowerCase();
    for (const charId of scene.characters_present) {
      const charName = validCharNames.get(charId);
      if (charName) {
        const firstName = charName.split(/\s+/)[0].toLowerCase();
        if (!imagePromptLower.includes(firstName) && !imagePromptLower.includes(charName.toLowerCase())) {
          // Soft notice (some prompts refer to generic roles or pronouns)
          issues.push(`Notice: Character name '${charName}' not explicitly mentioned in visual prompt text.`);
        }
      }
    }

    scene.consistency_issues = issues;
    totalIssuesCount += issues.length;
  }

  project.stage_outputs['stage_09'] = {
    total_scenes_checked: project.scenes.length,
    total_issues_flagged: totalIssuesCount,
    status: totalIssuesCount === 0 ? 'CLEAN_AND_CONSISTENT' : 'CONSISTENCY_NOTICES_RECORDED',
  };

  return project;
}
